java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -360,87 +335,37 @@ abstract class AbstractArcHost(
+        // TODO: add timeout?
+        val readyGate = Job()
+        val awaitingReady = particleContexts.map { it.particle }.toMutableSet()
+        val notifyReady: (Particle) -> Unit = {
+            if (awaitingReady.remove(it) && awaitingReady.isEmpty()) {
+                readyGate.complete()
             }
         }
 
-        if (context.arcState != ArcState.Error) {
-            setArcState(context, ArcState.Running)
-        }
-    }
+        // All particles have now received their startup events; move them to the running state.
+        particleContexts.forEach { it.runParticle(notifyReady) }
 
-    /**
-     * Invokes necessary [Particle] startup lifecycle methods given the current
-     * [ParticleContext.particleState], and changes that state as necessary.
-     */
-    private suspend fun performParticleLifecycle(arcId: String, particleContext: ParticleContext) {
-        val dispatcher = schedulerProvider(arcId).asCoroutineDispatcher()
-        if (particleContext.particleState == ParticleState.Instantiated) {
-            try {
-                withContext(dispatcher) {
-                    particleContext.particle.onFirstStart()
-                }
-                particleContext.particleState = ParticleState.FirstStart
-            } catch (e: Exception) {
-                log.error(e) { "Failure in particle during onFirstStart." }
-                markParticleAsFailed(particleContext)
-                return
-            }
-        }
-
-        // particleContext will take over the state handling after this.
-        try {
-            withContext(dispatcher) {
-                particleContext.particle.onStart()
-            }
-            particleContext.particleState = ParticleState.Waiting
-        } catch (e: Exception) {
-            log.error(e) { "Failure in particle during onStart." }
-            markParticleAsFailed(particleContext)
-            return
-        }
-    }
-
-    /**
-     * Move to [ParticleState.Failed] if this particle had previously successfully invoked
-     * [Particle.onFirstStart], else move to [ParticleState.Failed_NeverStarted]. Increments
-     * consecutive failure count, and if it reaches maximum, transitions to
-     * [ParticleState.MaxFailed].
-     *
-     * TODO: move into ParticleContext?
-     */
-    private fun markParticleAsFailed(particleContext: ParticleContext) {
-        particleContext.run {
-            if (particleState == ParticleState.MaxFailed) {
-                return
-            }
-
-            particleState = if (particleState.hasBeenStarted) {
-                ParticleState.Failed
-            } else {
-                ParticleState.Failed_NeverStarted
-            }
-            consecutiveFailureCount++
-
-            if (consecutiveFailureCount > MAX_CONSECUTIVE_FAILURES) {
-                particleState = ParticleState.MaxFailed
-            }
-        }
+        // Wait for the particles to get running.
+        readyGate.join()
```

Comment:
```
withTimeout(XXX) { 
   readyGate.join()
}
```

This will throw an exception, so someone has to catch it and set ArcState to Failed.